packages/vite/src/node/ssr/runtime/utils.ts

```diff
@@ -0,0 +1,276 @@
+import type { ImportMetaEnv } from './types'
+
+export const isWindows =
+  typeof process !== 'undefined' && process.platform === 'win32'
+
+export const decodeBase64 =
+  typeof atob !== 'undefined'
+    ? atob
+    : (str: string) => Buffer.from(str, 'base64').toString('utf-8')
+
+// currently we copy this from '../../constants' - maybe we can inline it somewhow?
+const NULL_BYTE_PLACEHOLDER = `__x00__`
+const VALID_ID_PREFIX = `/@id/`
+
+export function wrapId(id: string): string {
+  return id.startsWith(VALID_ID_PREFIX)
+    ? id
+    : VALID_ID_PREFIX + id.replace('\0', NULL_BYTE_PLACEHOLDER)
+}
+
+export function unwrapId(id: string): string {
+  return id.startsWith(VALID_ID_PREFIX)
+    ? id.slice(VALID_ID_PREFIX.length).replace(NULL_BYTE_PLACEHOLDER, '\0')
+    : id
+}
+
+const windowsSlashRE = /\\/g
+export function slash(p: string): string {
+  return p.replace(windowsSlashRE, '/')
+}
+
+export const queryRE = /\?.*$/s
+export const hashRE = /#.*$/s
+
+export function cleanUrl(url: string): string {
+  return url.replace(hashRE, '').replace(queryRE, '')
+}
+
+const _envShim = Object.create(null)
+
+const _getEnv = (environmentVariables?: Record<string, any>) =>
+  globalThis.process?.env ||
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore "env" in meta is not typed for SSR code
+  import.meta.env ||
+  // @ts-expect-error Deno global is not typed
+  globalThis.Deno?.env.toObject() ||
+  environmentVariables
```

Comment:
IIUC this correctly, this is used to proxy prop access to `import.meta.env.*` in the runtime? I'm not a big fan of this as it's not how we document `import.meta.env.*` to work. They are currently only loaded from `.env` files or `define`. Shouldn't Vite's transform flow already handle `import.meta.env.*` usage?

Reply:
Let's just remove it for now. Other runtimes can implement it themselves in their own custom `runner`